borsh-ts/test/serialize.test.js

```diff
@@ -0,0 +1,23 @@
+
+const borsh = require('../../borsh-js/index.js');
+
+class Test {
+    constructor(x, y, z, q) {
+        this.x = x;
+        this.y = y;
+        this.z = z;
+        this.q = q;
+    }
+}
+
+test('serialize object', async () => {
+    const value = new Test(255, 20, '123', [1, 2, 3]);
+    const schema = new Map([[Test, {kind: 'struct', fields: [['x', 'u8'], ['y', 'u64'], ['z', 'string'], ['q', [3]]] }]]);
```

Comment:
We could have a derive macro in Rust that generates this spec for the decorated types. We then can use this macro to generate ABI part of the contact metadata: https://github.com/nearprotocol/NEPs/pull/3

Reply:
This is not contract metadata though, this is SDK for NEAR client.

Reply:
Is there a need to introduce ABI?

Reply:
> This is not contract metadata though, this is SDK for NEAR client.

Yes, I understand. My comment might be confusing, because it talks both about generating spec using rust macro and ABI.

Regarding the first part: Currently when someone wants to serialize a JS object they need to write a schema by hand. Ideally we would like this schema to be generatable just like Protobuffs generate bindings for different languages. With protobufs one can feed .proto files describing structs to a generator tool that would produce bindings in JS. In borsh case we can have Rust structs (we don't need a separate language like proto to describe objects) in a crate and generate JS schema from them using macroses and other Rust tools.

> And as far as I understand we are no planning to introduce ABIs in nearprotocol/NEPs#3 @vgrichina @bowenwang1996

Let's move this discussion to nearprotocol/NEPs#3 but my understanding is that it is mostly about ABI, plus some other static metadata, see https://github.com/nearprotocol/NEPs/pull/3/files#diff-12749e075bd05517de4a2627d3167a23R177